ffi_internals/src/consumer.rs

```diff
@@ -109,6 +113,47 @@ fn write_primitive_conformances(consumer_dir: &str) -> Result<(), std::io::Error
     })
 }
 
+/// Describes the behaviors required for generating a consumer type.
+///
+pub trait ConsumerType {
+    /// The type definition.
+    ///
+    fn type_definition(&self) -> String;
+
+    /// The implementation of the `NativeData` protocol.
+    fn native_data_impl(&self) -> String;
+
+    /// The implementation of the `FFIArray` protocol.
+    ///
+    fn ffi_array_impl(&self) -> String;
+
+    /// The implementation of the `NativeArrayData` protocol.
+    ///
+    fn native_array_data_impl(&self) -> String;
+
+    /// The implementation for handling optional instances of this type.
+    ///
+    fn option_impl(&self) -> String;
+
+    /// Additional imports for the consumer side that this type requires.
+    ///
+    fn required_imports(&self) -> &[syn::Path];
+}
+
+impl From<&dyn ConsumerType> for String {
```

Comment:
Ah... yeah... does this have to be a trait object? It might make things a bit easier to work with if you

```rust
impl<C> From<&C> for String
where
    C: ConsumerType
{
}
```

...or maybe you could even go a different route and:

```rust
impl<T> std::fmt::Display for T
where
    T: ConsumerType
{
}
```

I suppose at this point it's really just semantics, but I thought I'd throw them out there in case it works.

Reply:
Derp. Yeah, a generic impl constrained to `Consumer` makes more sense and will get rid of the weird casting above. 🤦 
Thanks!